### PR TITLE
Jenkins plugin instance type search

### PR DESCRIPTION
**Pull request by Lironrad**

### CODE REVIEW

File: src/main/java/hudson/plugins/spotinst/cloud/SpotinstInstanceWeight.java [hunk 1]
```diff
@@ -49,6 +68,57 @@ public Descriptor<SpotinstInstanceWeight> getDescriptor() {
     }
     //endregion
 
+    //region Methods
+    public String getAwsInstanceTypeFromAPI() {
+        String type = null;
+        switch(getSearchMethod()){
```
Comment (Lironrad) [inline src/main/java/hudson/plugins/spotinst/cloud/SpotinstInstanceWeight.java]: coding conventions: extract to a veritable outside from the switch statement:

AwsInstanceTypeSearchMethodEnum searchMethod = getSearchMethod()
switch(searchMethod)

File: src/main/java/hudson/plugins/spotinst/cloud/SpotinstInstanceWeight.java [hunk 2]
```diff
@@ -71,13 +128,41 @@ public ListBoxModel doFillAwsInstanceTypeFromAPIItems() {
             return retVal;
         }
 
-        public FormValidation doCheckAwsInstanceTypeFromAPI() {
+        public AutoCompletionCandidates doAutoCompleteAwsInstanceTypeFromAPISearch(@QueryParameter String value) {
+            AutoCompletionCandidates c = new AutoCompletionCandidates();
```
Comment (Lironrad) [inline src/main/java/hudson/plugins/spotinst/cloud/SpotinstInstanceWeight.java]: coding conventions, change to: AutoCompletionCandidates retVal = new AutoCompletionCandidates();

File: src/main/java/hudson/plugins/spotinst/cloud/SpotinstInstanceWeight.java [hunk 3]
```diff
@@ -49,6 +55,58 @@ public Descriptor<SpotinstInstanceWeight> getDescriptor() {
...
 
+    //region Methods
+    public String getAwsInstanceTypeFromAPI() {
+        String type = null;
+        AwsInstanceTypeSearchMethodEnum searchMethod = getSearchMethod();
+        switch(searchMethod){
+            case SEARCH:
+                type = getAwsInstanceTypeFromAPISearch();
+                break;
+            case SELECT:
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/SpotinstInstanceWeight.java]: Remove the default

File: src/main/java/hudson/plugins/spotinst/cloud/SpotinstInstanceWeight.java [hunk 4]
```diff
@@ -71,13 +129,41 @@ public ListBoxModel doFillAwsInstanceTypeFromAPIItems() {
             return retVal;
         }
 
-        public FormValidation doCheckAwsInstanceTypeFromAPI() {
+        public AutoCompletionCandidates doAutoCompleteAwsInstanceTypeFromAPISearch(@QueryParameter String value) {
+            AutoCompletionCandidates retVal = new AutoCompletionCandidates();
+            List<AwsInstanceType> allAwsInstanceTypes = SpotAwsInstanceTypesHelper.getAllInstanceTypes();
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/SpotinstInstanceWeight.java]: Dont run the api for each letter, use the existing item

File: src/main/java/hudson/plugins/spotinst/cloud/SpotinstInstanceWeight.java [hunk 5]
```diff
@@ -71,13 +129,41 @@ public ListBoxModel doFillAwsInstanceTypeFromAPIItems() {
...
+            Stream<String> allTypes      = allAwsInstanceTypes.stream()
+                                                              .map(awsInstanceType -> awsInstanceType.getInstanceType().toLowerCase());
+            String filterValue = value.toLowerCase();
+            Object[] matchingTypes = allTypes.filter(type -> type.startsWith(filterValue)).toArray();
+
+            for (Object objectType: matchingTypes) {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/SpotinstInstanceWeight.java]: use stream.foreach

**Review by zivmessing — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/spotinst/cloud/SpotinstInstanceWeight.java [hunk 6]
```diff
@@ -23,9 +27,11 @@
 public class SpotinstInstanceWeight implements Describable<SpotinstInstanceWeight> {
     //region Members
     private Integer             executors;
-    private String              awsInstanceTypeFromAPI;
+    private String              awsInstanceTypeFromAPISelect;
+    private String              awsInstanceTypeFromAPISearch;
+    private AwsInstanceTypeSearchMethodEnum searchMethod;
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/SpotinstInstanceWeight.java]: Chaeck also if we can add error message for invalid instance type

**Review by zivmessing — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/spotinst/common/AwsInstanceTypeSearchMethodEnum.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
+package hudson.plugins.spotinst.common;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public enum AwsInstanceTypeSearchMethodEnum {
+    SELECT("SELECT"),
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/common/AwsInstanceTypeSearchMethodEnum.java]: change to PICK

File: src/main/java/hudson/plugins/spotinst/common/AwsInstanceTypeSearchMethodEnum.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
+package hudson.plugins.spotinst.common;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public enum AwsInstanceTypeSearchMethodEnum {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/common/AwsInstanceTypeSearchMethodEnum.java]: rename to AwsInstanceTypeSelectMethodEnum

File: src/main/java/hudson/plugins/spotinst/common/AwsInstanceTypeSearchMethodEnum.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+    AwsInstanceTypeSearchMethodEnum(String name) {
+        this.name = name;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public static AwsInstanceTypeSearchMethodEnum fromValue(String value) {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/common/AwsInstanceTypeSearchMethodEnum.java]: rename to fromName

File: src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    @Override
+    public String getDisplayName() {
+        return "Aws Spotinst Cloud Instance Type Monitor";
+    }
+    //endregion
+
+    //region getters & setters
+    public boolean hasInvalidInstanceType() {
+        List<String> cachedInvalidInstances = getCachedInvalidInstance();
+        boolean retVal = cachedInvalidInstances != null && cachedInvalidInstances.isEmpty() == false;
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java]: Call : Use findInvalidInstancesFromClouds and check if empty by collectionUtils.isEmpty(cachedInvalidInstances)

File: src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+        awsClouds.forEach(awsCloud -> {
+            List<String> invalidTypes = awsCloud.getInvalidInstanceTypes();
+            retVal.addAll(invalidTypes);
+        });
+
+        return retVal;
+    }
+
+    private List<String> getCachedInvalidInstance() {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java]: not needed , only one line

File: src/main/java/hudson/plugins/spotinst/common/SpotinstContext.java [hunk 12]
```diff
@@ -63,6 +65,18 @@ public Date getAwsInstanceTypesLastUpdate() {
     public void setAwsInstanceTypesLastUpdate(Date awsInstanceTypesLastUpdate) {
         this.awsInstanceTypesLastUpdate = awsInstanceTypesLastUpdate;
     }
+
+    public List<String> getInvalidInstanceTypes() {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/common/SpotinstContext.java]: not needed as cache

File: src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    List<String> invalidInstances;
+    //endregion
+
+    //region Overrides
+    @Override
+    public boolean isActivated() {
+        boolean     retVal                = false;
+        Set<String> allCloudsInvalidTypes = findInvalidInstancesFromClouds();
+
+        SpotinstContext.getInstance().setInvalidInstanceTypes(new ArrayList<>(allCloudsInvalidTypes));
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java:31]: just return true - try it

File: src/main/resources/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor/message.properties [hunk 14]
```diff
@@ -0,0 +1,3 @@
+Title=Invalid AWS Instance types
+InvalidInstanceType=Cloud has settings for invalid instance types <i>Types</i>: <i>{0}</i>
+Explanation= Please use another Instance type or delete the types' configuration from the cloud configuration.
```
Comment (zivmessing) [inline src/main/resources/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor/message.properties:3]: Add the group id also

**Review by zivmessing — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java (hunk 13 above)
Comment (sitay93) [inline src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java]: Put always an empty alert. should do the check

File: src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Extension
+public class AwsSpotinstCloudInstanceTypeMonitor extends AdministrativeMonitor {
+    //region members
+    private static final Logger LOGGER = LoggerFactory.getLogger(AwsSpotinstCloud.class);
+    Map<String, List<String>> invalidInstances;
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java]: rename to invalidInstancesByGroupId

File: src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    //endregion
+
+    //region Overrides
+    @Override
+    public boolean isActivated() {
+        boolean retVal;
+        invalidInstances = new HashMap<>();
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java]: extract to method : initinvalidInstances

**Review by zivmessing — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+        awsClouds.forEach(awsCloud -> {
+            String       elastigroupId = awsCloud.getGroupId();
+            List<String> invalidTypes  = awsCloud.getInvalidInstanceTypes();
+
+            if (CollectionUtils.isEmpty(invalidTypes) == false) {
+                invalidInstances.put(elastigroupId, invalidTypes);
+            }
+        });
+
+        retVal = hasInvalidInstanceType();
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java:25]: if retVal == true (has invalid) generate the alert message and print it to log

File: src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java (hunk 15 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java]: Done

File: src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java (hunk 16 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/monitor/AwsSpotinstCloudInstanceTypeMonitor.java]: Done

**Review by zivmessing — APPROVED** [review]

**Review by adivos — APPROVED** [review]

**Pull request merged**